trl/trainer/online_dpo_trainer.py

```diff
@@ -129,15 +133,40 @@ def __init__(
         if tokenizer is None:
             raise ValueError("`tokenizer` must be provided.")
 
-        # We don't optimize the reward model model nor the ref model, so we can set them to eval mode
-        self.ref_model.eval()
-        if self.reward_model is not None:
-            self.reward_model.eval()
+        # Convert to PEFT model if peft_config is provided
+        if peft_config is not None:
+            # Check if PEFT is available
+            if not is_peft_available():
+                raise ImportError(
+                    "PEFT is not available and passed `peft_config`. Please install PEFT with "
+                    "`pip install peft` to use it."
+                )
+
+            # Get peft model with the given config
+            model = get_peft_model(model, peft_config)
 
         # Disable dropout in the model if specified
         if args.disable_dropout:
             disable_dropout_in_model(model)
 
+        # Handle the ref_model
+        # Usually, the user wants the ref model to be the initial version of the model. When using PEFT, it's easy to
+        # get the ref model, as it's just the model with a disabled adaptater. When not using PEFT, we need to create
+        # the ref model from the model by copying it and disable the gradients and set it in evaluation mode.
+        if ref_model is None:  # No ref model provided, the most common case
+            if peft_config is None:
+                self.ref_model = create_reference_model(model)  # copy, disable gradients, set eval mode
+            else:
+                self.ref_model = None  # we don't need a ref model here, we can just disable the adaptater.
+        else:  # rare case, the user provided a ref model
+            self.ref_model = create_reference_model(ref_model)  # copy, disable gradients, set eval mode
+
+        # Disable the gradient and set the reward model in eval mode
+        if self.reward_model is not None:
+            self.reward_model.eval()
+            for param in self.reward_model.parameters():
+                param.requires_grad = False
```

Comment:
I am not sure why this is needed, as surely it is called with a `no_grad` context?

Reply:
eval is needed.

this

```python
for param in self.reward_model.parameters():
    param.requires_grad = False
...
model(...)
```

is the same as 

```python
with torch.no_grad():
    model(...)
```

do you think the later makes more sense?

Reply:
Let's opt for the second option, it is preferable to limit the modification of input objects